### PR TITLE
Overarching PR for features related to generating code starting from a parameter.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/15648

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide This is ready to be reviewed.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/analyzer-ioperation as this required a small change to IOperation to work. Specifically, i had to make it so that you coudl get the IBlockStatement given a VB MethodBlockBaseSyntax.

Tagging @dotnet/roslyn-compiler as the analyzer change is in compiler code.  I assume compiler sign-off rules still apply even for code that is entirely IOperation related.

File: src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb [hunk 1]
```diff
@@ -351,7 +351,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+            Dim methodBlock = TryCast(node, MethodBlockBaseSyntax)
+            If methodBlock IsNot Nothing Then
+                ' Trying to get the MemberSemanticModel for a MethodBlock will end up returning
+                ' nothing.  That's because trying to get Binder for the MethodBlock will actually
+                ' return the binder for the containing type.  To avoid this we ask for the binder
+                ' for the methodBlock, passing in a position at the end of it's starting block-statement.
+                ' This will cause it to try to get the interior MethodBodyBinder which we can then
+                ' effectively get a MemberSemanticModel from.
+                Dim binder = _binderFactory.GetBinderForPosition(methodBlock, methodBlock.BlockStatement.EndPosition)
+                model = GetMemberSemanticModel(binder)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb]: I think we should use ```Friend Function GetMemberSemanticModel(position As Integer) As MemberSemanticModel``` instead.

File: src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb [hunk 2]
```diff
@@ -351,7 +351,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
         Friend Overrides Function GetOperationWorker(node As VisualBasicSyntaxNode, options As GetOperationOptions, cancellationToken As CancellationToken) As IOperation
-            Dim model As MemberSemanticModel = Me.GetMemberSemanticModel(node)
+            Dim model As MemberSemanticModel
+
+            Dim methodBlock = TryCast(node, MethodBlockBaseSyntax)
+            If methodBlock IsNot Nothing Then
+                ' Trying to get the MemberSemanticModel for a MethodBlock will end up returning
+                ' nothing.  That's because trying to get Binder for the MethodBlock will actually
+                ' return the binder for the containing type.  To avoid this we ask for the binder
+                ' for the methodBlock, passing in a position at the end of it's starting block-statement.
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb]: nit: typo "its"

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+interface I
+{
+    void M([||]string s);
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInitializeParameter)]
+        public async Task TestNotOnAbstractParameter()
+        {
+            await TestMissingInRegularAndScriptAsync(
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: Added.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+    {
+        if (s == null)
+        {
+            throw new ArgumentNullException(nameof(s));
+        }
+        
+        Init();
+    }
+}");
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: added.

File: src/EditorFeatures/CSharpTest/InitializeParameter/InitializeMemberFromParameterTests.cs [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+
+    public C(string s)
+    {
+        S1 = s;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInitializeParameter)]
+        public async Task TestInitializePropertyWithBadName()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/InitializeMemberFromParameterTests.cs]: I'm just validating that we don't hook up to a property with a name that certainly doesn't match.  will rename test.

File: src/EditorFeatures/CSharpTest/InitializeParameter/InitializeMemberFromParameterTests.cs [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+class C
+{
+    private int s;
+
+    public C([||]string s)
+    {
+        s = 0;
+    }
+}",
+
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/InitializeMemberFromParameterTests.cs]: Fixed.

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs Are you ok with teh compiler side with the change to call GetMemberSemanticModel?

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs could you let me know about the IOperation bit?  Thanks!

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler Can you take a look?  Thanks!

File: src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb [hunk 7]
```diff
@@ -351,7 +351,19 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+            Dim model As MemberSemanticModel
+
+            Dim methodBlock = TryCast(node, MethodBlockBaseSyntax)
+            If methodBlock IsNot Nothing Then
+                ' Trying to get the MemberSemanticModel for a MethodBlock will end up returning
+                ' nothing.  That's because trying to get Binder for the MethodBlock will actually
+                ' return the binder for the containing type.  To avoid this we ask for the model
+                ' passing in a position at the end of the method's starting block-statement.
+                ' This will cause it to try to get the interior MemberSemanticModel.
+                model = GetMemberSemanticModel(methodBlock.BlockStatement.EndPosition)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb]: >model = GetMemberSemanticModel(methodBlock.BlockStatement.EndPosition) [](start = 16, length = 70)

I think we should have  targeted IOperation tests for this code path for all different kinds of syntax nodes that can get into this ```if```, including abstract declarations.

Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb]: Sure.  I'll make sure the IOperation test plan includes that.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/SyntaxTreeSemanticModel.vb]: > Sure. I'll make sure the IOperation test plan includes that.

If you plan to merge this PR, the tests should be included.

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs I've added some tests for several MethodBlockBaseSyntax cases.  Let me know if there are more that you'd like me to add.  Note: while doing this testing i opened https://github.com/dotnet/roslyn/issues/18417 as an interesting design question we'll likely want to look more into.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplayVisitor.vb [hunk 8]
```diff
@@ -239,7 +239,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         End Sub
 
         Public Overrides Sub VisitLocal(symbol As ILocalSymbol)
-            builder.Add(CreatePart(SymbolDisplayPartKind.LocalName, symbol, symbol.Name, False))
+            If symbol.Name IsNot Nothing Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplayVisitor.vb]: >symbol.Name IsNot Nothing [](start = 15, length = 25)

It feels like we should output something if this condition is false.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        If 1 > 2
+        End If
+
+        Return 0;
+    End Operator
+End Class"
+
+            Dim expectedOperationTree = "
+IBlockStatement (4 statements, 1 locals) (OperationKind.BlockStatement)
+  Local_1:  As System.Int32
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb]: > As System.Int32 [](start = 11, length = 16)

This symbol display doesn't feel right. Perhaps we should output ```<anonymous>``` as a replacement for the name.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    IBlockStatement (0 statements) (OperationKind.BlockStatement)
+  ILabelStatement (Label: exit) (OperationKind.LabelStatement)
+  IReturnStatement (OperationKind.ReturnStatement)
+    ILocalReferenceExpression: Prop (OperationKind.LocalReferenceExpression, Type: System.Int32)"
+
+            VerifyOperationTreeForTest(Of MethodBlockBaseSyntax)(source, expectedOperationTree)
+        End Sub
+
+        <Fact>
+        Public Sub IBlockStatement_PropertySetBlock()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb]: >IBlockStatement_PropertySetBlock [](start = 19, length = 32)

Consider also testing event acessors.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb (hunk 9 above)
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb]: Sounds good.  #fixed.

File: src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplayVisitor.vb (hunk 8 above)
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/SymbolDisplay/SymbolDisplayVisitor.vb]: #fixed

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb (hunk 10 above)
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb]: #done.

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs Latest feedback responded to.  Please look when possible.

Also @jaredpar What are your thoughts on these PRs and minor IOperation changes in the compiler?  Does it suffice to have a single Compiler signoff and one IOperation signoff?  Or are two compiler sign-offs still necessary?

Thanks!

**Comment by AlekseyTs** [comment]
CC @dotnet/roslyn-compiler For the second sign-off.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler While this PR looks large, the actual compiler side is quite tiny (just a couple of tweaked methods, and some test coverage).  If someone can take a look asap, i would be very appreciative.  Thanks!

**Pull request merged**